File: app/src/main/java/com/example/semp/FazerPedidoActivity.java
package com.example.semp;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.util.Patterns;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Spinner;
import android.widget.Toast;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.util.Pair;
import androidx.core.view.GravityCompat;
import androidx.drawerlayout.widget.DrawerLayout;

import com.example.semp.models.GenericResponse;
import com.example.semp.models.PedidoRequest;
import com.example.semp.models.Produto;
import com.google.android.material.datepicker.MaterialDatePicker;
import com.google.gson.Gson;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Locale;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class FazerPedidoActivity extends AppCompatActivity {

    private EditText etEmail, etDataReserva, etJustificativa, etNomeUsuario, etCodigoPedido;
    private Spinner spinnerPrioridade;
    private Button btnConfirmar, btnGerarCodigo;
    private List<Produto> listaProdutosParaPedido = new ArrayList<>();
    private DrawerLayout drawerLayout;

    // Variáveis seguras da sessão
    private String usuarioSeguro = "Usuario";
    private String unidadeSegura = "Unidade Central";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_fazer_pedido);

        drawerLayout = findViewById(R.id.drawerLayoutFazerPedido);
        View btnMenu = findViewById(R.id.btnMenu);
        if (btnMenu != null) {
            btnMenu.setOnClickListener(v -> {
                if (drawerLayout != null) drawerLayout.openDrawer(GravityCompat.START);
            });
        }
        MenuSidebarHelper.configurarNavegacao(this, drawerLayout);

        // Busca dados da sessão de forma segura (previne NullPointerException)
        SharedPreferences prefs = getSharedPreferences("SessaoApp", Context.MODE_PRIVATE);
        usuarioSeguro = prefs.getString("usuarioLogado", "Usuario");
        unidadeSegura = prefs.getString("unidadeAtual", "Unidade Central");

        if (getIntent().hasExtra("ITENS_SELECIONADOS")) {
            String json = getIntent().getStringExtra("ITENS_SELECIONADOS");
            if (json != null) {
                java.lang.reflect.Type listType = new com.google.gson.reflect.TypeToken<List<Produto>>() {}.getType();
                listaProdutosParaPedido = new Gson().fromJson(json, listType);
            }
        }

        etNomeUsuario = findViewById(R.id.etNomeUsuario);
        etEmail = findViewById(R.id.etEmail);
        etDataReserva = findViewById(R.id.etDataReserva);
        etJustificativa = findViewById(R.id.etJustificativa);
        etCodigoPedido = findViewById(R.id.etCodigoPedido);
        spinnerPrioridade = findViewById(R.id.spinnerPrioridade);
        btnConfirmar = findViewById(R.id.btnConfirmarEmprestimo);
        btnGerarCodigo = findViewById(R.id.btnGerarCodigoPedido);
        View btnVoltar = findViewById(R.id.btnVoltarCarrinho);

        if (etNomeUsuario != null) {
            etNomeUsuario.setText(usuarioSeguro);
        }

        if (spinnerPrioridade != null) {
            String[] prioridades = {"Baixo", "Médio", "Alto"};
            ArrayAdapter<String> adapter = new ArrayAdapter<>(this, android.R.layout.simple_spinner_item, prioridades);
            adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
            spinnerPrioridade.setAdapter(adapter);
        }

        if (etDataReserva != null) {
            etDataReserva.setOnClickListener(v -> abrirCalendario());
        }

        if (btnGerarCodigo != null) {
            btnGerarCodigo.setOnClickListener(v -> {
                String idEstado = prefs.getString("id_estado", "X");
                String idRegiao = prefs.getString("id_regiao", "X");
                int idUnidade = prefs.getInt("id_unidade", 0);
                
                String novoCodigo = SempUtils.gerarCodigoPedidoModerno(idEstado, idRegiao, idUnidade);
                etCodigoPedido.setText(novoCodigo);
                Toast.makeText(this, "Código do Pedido Gerado!", Toast.LENGTH_SHORT).show();
            });
        }

        if (btnVoltar != null) {
            btnVoltar.setOnClickListener(v -> finish());
        }

        btnConfirmar.setOnClickListener(v -> efetivarPedido());
    }

    private void abrirCalendario() {
        MaterialDatePicker.Builder<Pair<Long, Long>> builder = MaterialDatePicker.Builder.dateRangePicker();
        builder.setTitleText("Selecione o período da reserva");

        final MaterialDatePicker<Pair<Long, Long>> picker = builder.build();
        picker.addOnPositiveButtonClickListener(selection -> {
            if (selection.first != null && selection.second != null) {
                SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd", Locale.getDefault());
                String dataInicio = sdf.format(new Date(selection.first));
                String dataFim = sdf.format(new Date(selection.second));
                
                String periodo = dataInicio + " até " + dataFim;
                etDataReserva.setText(periodo);
            }
        });
        picker.show(getSupportFragmentManager(), "RANGE_PICKER");
    }

    private void efetivarPedido() {
        if (listaProdutosParaPedido == null || listaProdutosParaPedido.isEmpty()) {
            Toast.makeText(this, "Nenhum produto selecionado no carrinho!", Toast.LENGTH_SHORT).show();
            return;
        }

        String email = etEmail.getText().toString().trim();
        String dataReserva = etDataReserva.getText().toString().trim();
        String justificativa = etJustificativa.getText().toString().trim();

        if (email.isEmpty() || !Patterns.EMAIL_ADDRESS.matcher(email).matches()) {
            etEmail.setError("Insira um e-mail válido");
            etEmail.requestFocus();
            return;
        }

        if (dataReserva.isEmpty()) {
            etDataReserva.setError("Selecione uma data");
            return;
        }

        btnConfirmar.setEnabled(false);
        btnConfirmar.setText("Enviando...");

        String dataPostagemComHora = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault()).format(new Date());

        String prioridadeSelecionada = spinnerPrioridade.getSelectedItem() != null ? spinnerPrioridade.getSelectedItem().toString() : "Baixo";
        String prioridadeParaDB = "baixo";
        switch (prioridadeSelecionada) {
            case "Alto": prioridadeParaDB = "alto"; break;
            case "Médio": prioridadeParaDB = "intermediário"; break;
        }

        // GERA O CÓDIGO DO PEDIDO (SE NÃO TIVER SIDO GERADO MANUALMENTE)
        String codigoPedidoGerado = etCodigoPedido.getText().toString().trim();
        if (codigoPedidoGerado.isEmpty()) {
            Toast.makeText(this, "Por favor, gere o código do pedido antes de confirmar!", Toast.LENGTH_SHORT).show();
            return;
        }

        // MONTA A LISTA DE PRODUTOS COM DADOS REAIS
        List<PedidoRequest.ProdutoPedido> produtosFormatados = new ArrayList<>();
        SharedPreferences sessao = getSharedPreferences("SessaoApp", Context.MODE_PRIVATE);
        for (Produto p : listaProdutosParaPedido) {
            // Usa o código do produto que já existe ou gera um novo prefixado com 2
            String idEst = sessao.getString("id_estado", "X");
            String idReg = sessao.getString("id_regiao", "X");
            int idUni = sessao.getInt("id_unidade", 0);
            String codigoProdutoGerado = SempUtils.gerarCodigoProdutoModerno(idEst, idReg, idUni);
            
            // Pega a quantidade correta do carrinho
            int qtdSolicitada = p.quantidade > 0 ? p.quantidade : (p.carrinho > 0 ? p.carrinho : 1);

            produtosFormatados.add(new PedidoRequest.ProdutoPedido(
                    p.codigo != null ? p.codigo : codigoProdutoGerado,
                    p.nome != null ? p.nome : "Produto Sem Nome",
                    qtdSolicitada,
                    p.unidade_atual != null ? p.unidade_atual : unidadeSegura,
                    codigoPedidoGerado
            ));
        }

        PedidoRequest request = new PedidoRequest(
                usuarioSeguro, email, unidadeSegura, dataReserva,
                produtosFormatados, prioridadeParaDB,
                justificativa.isEmpty() ? "Solicitação de Empréstimo" : justificativa,
                dataPostagemComHora, codigoPedidoGerado
        );

        RetrofitClient.getApi().fazerPedido(request).enqueue(new Callback<GenericResponse>() {
            @Override
            public void onResponse(Call<GenericResponse> call, Response<GenericResponse> response) {
                btnConfirmar.setEnabled(true);
                btnConfirmar.setText("Confirmar Pedido");
                if (response.isSuccessful() && response.body() != null) {
                    if (response.body().sucesso) {
                        Toast.makeText(FazerPedidoActivity.this, "Pedido efetivado! Cód: " + codigoPedidoGerado, Toast.LENGTH_LONG).show();
                        finish();
                    } else {
                        // EXIBE A MENSAGEM DE CONFLITO VINDA DO SERVIDOR
                        String msgErro = response.body().mensagem != null ? response.body().mensagem : "Erro ao processar reserva.";
                        Toast.makeText(FazerPedidoActivity.this, msgErro, Toast.LENGTH_LONG).show();
                    }
                } else {
                    Toast.makeText(FazerPedidoActivity.this, "Erro no servidor: " + response.code(), Toast.LENGTH_LONG).show();
                }
            }

            @Override
            public void onFailure(Call<GenericResponse> call, Throwable t) {
                btnConfirmar.setEnabled(true);
                btnConfirmar.setText("Confirmar Pedido");
                Toast.makeText(FazerPedidoActivity.this, "Erro de conexão", Toast.LENGTH_SHORT).show();
            }
        });
    }
}

File: app/src/main/java/com/example/semp/FazerPedidoActivity.kt
package com.example.semp

import android.app.DatePickerDialog
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.EditText
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.GravityCompat
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.drawerlayout.widget.DrawerLayout
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.util.Calendar

class FazerPedidoActivity : AppCompatActivity() {

    private var drawerLayout: DrawerLayout? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        try {
            setContentView(R.layout.activity_fazer_pedido)

            drawerLayout = findViewById(R.id.drawerLayoutPedido) ?: findViewById(R.id.drawerLayout)
            val btnMenu = findViewById<ImageView>(R.id.btnMenu)
            val etNome = findViewById<EditText>(R.id.etNomePedido)
            val etEmail = findViewById<EditText>(R.id.etEmailPedido)
            val etData = findViewById<EditText>(R.id.etDataPedido)
            val btnConfirmar = findViewById<Button>(R.id.btnConfirmarPedido)

            findViewById<View>(R.id.mainContentLayout)?.let { mainView ->
                ViewCompat.setOnApplyWindowInsetsListener(mainView) { view, insets ->
                    val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
                    view.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
                    insets
                }
            }

            btnMenu?.setOnClickListener { drawerLayout?.openDrawer(GravityCompat.START) }
            configurarNavegacaoMenu()

            etData?.setOnClickListener {
                val calendar = Calendar.getInstance()
                val year = calendar.get(Calendar.YEAR)
                val month = calendar.get(Calendar.MONTH)
                val day = calendar.get(Calendar.DAY_OF_MONTH)

                DatePickerDialog(this, { _, selectedYear, selectedMonth, selectedDay ->
                    val dataFormatada = "$selectedYear-${selectedMonth + 1}-$selectedDay"
                    etData.setText(dataFormatada)
                }, year, month, day).show()
            }

            btnConfirmar?.setOnClickListener {
                val nome = etNome?.text.toString()
                val email = etEmail?.text.toString()
                val data = etData?.text.toString()
                val unidade = MainActivity.getUnidade()

                if (nome.isEmpty() || email.isEmpty() || data.isEmpty()) {
                    Toast.makeText(this, "Preencha todos os campos", Toast.LENGTH_SHORT).show()
                    return@setOnClickListener
                }

                enviarPedidoParaAPI(nome, email, unidade, data, etNome, etEmail, etData)
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private fun enviarPedidoParaAPI(nome: String, email: String, unidade: String, data: String, etNome: EditText?, etEmail: EditText?, etData: EditText?) {
        val request = PedidoRequest(nome, email, unidade, data)

        RetrofitClient.api.fazerPedido(request).enqueue(object : Callback<GenericResponse> {
            override fun onResponse(call: Call<GenericResponse>, response: Response<GenericResponse>) {
                if (isDestroyed || isFinishing) return // Bloqueia crash se a tela foi fechada

                val body = response.body()
                if (response.isSuccessful && body != null && body.sucesso) {
                    Toast.makeText(this@FazerPedidoActivity, "Dados inseridos com sucesso!", Toast.LENGTH_SHORT).show()
                    etNome?.text?.clear()
                    etEmail?.text?.clear()
                    etData?.text?.clear()
                } else {
                    Toast.makeText(this@FazerPedidoActivity, body?.mensagem ?: "Erro ao registrar o pedido", Toast.LENGTH_LONG).show()
                }
            }

            override fun onFailure(call: Call<GenericResponse>, t: Throwable) {
                if (!isDestroyed && !isFinishing) {
                    Toast.makeText(this@FazerPedidoActivity, "Erro de conexão com servidor", Toast.LENGTH_LONG).show()
                }
            }
        })
    }

    private fun configurarNavegacaoMenu() {
        val nivel = MainActivity.getNivelConta()

        val btnConfigEstoque = findViewById<TextView>(R.id.menuItemConfigEstoque)
        val btnAutorizar = findViewById<TextView>(R.id.menuItemAutorizar)
        val btnConfigAcesso = findViewById<TextView>(R.id.menuItemConfigAcesso)
        val btnEmprestimo = findViewById<TextView>(R.id.menuItemEmprestimo)
        val btnVisualizarPedido = findViewById<TextView>(R.id.menuItemVisualizarPedido)

        // Limita o acesso dependendo da conta do usuário
        btnConfigEstoque?.visibility = if (nivel == "1" || nivel == "2") View.VISIBLE else View.GONE
        btnAutorizar?.visibility = if (nivel == "2") View.VISIBLE else View.GONE
        btnConfigAcesso?.visibility = if (nivel == "1") View.VISIBLE else View.GONE

        findViewById<TextView>(R.id.menuItemEstoque)?.setOnClickListener { startActivity(Intent(this, EstoqueActivity::class.java)); finish() }
        findViewById<TextView>(R.id.menuItemCarrinho)?.setOnClickListener { startActivity(Intent(this, CarrinhoActivity::class.java)); finish() }

        btnConfigEstoque?.setOnClickListener { startActivity(Intent(this, ConfigEstoqueActivity::class.java)); finish() }
        btnAutorizar?.setOnClickListener { startActivity(Intent(this, AutorizarPedidosActivity::class.java)); finish() }
        btnConfigAcesso?.setOnClickListener { startActivity(Intent(this, CadastrarUsuarioActivity::class.java)); finish() }

        btnEmprestimo?.setOnClickListener { Toast.makeText(this, "Empréstimo em breve", Toast.LENGTH_SHORT).show() }
        btnVisualizarPedido?.setOnClickListener { Toast.makeText(this, "Visualização de Pedidos em breve", Toast.LENGTH_SHORT).show() }

        findViewById<TextView>(R.id.menuItemSair)?.setOnClickListener {
            startActivity(Intent(this, MainActivity::class.java).apply { flags = Intent.FLAG_ACTIVITY_CLEAR_TOP }); finish()
        }
    }
}